iroh/src/node/rpc.rs

```diff
@@ -990,50 +970,94 @@ impl<D: BaoStore> Handler<D> {
     }
 }
 
-async fn download_and_export<D, C, F>(
-    db: D,
-    get_conn: C,
-    hash_and_format: HashAndFormat,
-    out: DownloadLocation,
-    tag: SetTagOption,
-    progress: impl ProgressSender<Msg = DownloadProgress> + IdGenerator,
+async fn download<D>(
+    db: &D,
+    endpoint: MagicEndpoint,
+    downloader: &Downloader,
+    req: BlobDownloadRequest,
+    progress: FlumeProgressSender<DownloadProgress>,
```

Comment:
To avoid this whole id mapping thing, couldn't you have a global id generator? Then the ids for different downloads would never overlap, and you could join them without having to do the id mapping...

Reply:
This is what I had until a couple of commits ago. However! After the download finishes, the export runs, outside of the scope of the downloader. See the changes in `iroh/src/node/rpc.rs`.
So the flow roughly like this:
```rust
let progress =  FlumeProgressSender::new();
let handle = downloader.queue(progress.clone());
handle.await?;
progress.send(NetworkDone).await;
export_download(progress.clone()).await?;
progress.send(AllDone).await;
```
the `export_download` again creates ids with `progress.new_id()`. Without the id mapping, these progress ids could be the same ones that were emitted during the download.

I couldn't come up with a nice way to solve this, that's why I resorted to the id map. I agree that it's not too nice, but it is the most flexible if we want to keep the current premise of "just pass in any progress sender".

I'm open to other ideas...

Reply:
With #2113 a download now only gets progress IDs from within the downloader, so we can remove the whole id mapping foo!